internal/planner/planner.go

```diff
@@ -452,6 +457,8 @@ func (p *Planner) planQuery(q ast.Body, index int, iter planiter) error {
 
 // TODO(tsandall): improve errors to include location information.
 func (p *Planner) planExpr(e *ast.Expr, iter planiter) error {
+	p.setLocation(e.Location)
```

Comment:
Yeah I was thinking about that problem. I hadn't been able to trigger it, but I'll have another look! Thanks.

Reply:
Here's an example where the problem is observable.

The locations for statements in the function body are essentially correct (I'm assuming 0:1:1 is some empty/default value). The last three statements are at "0:6:4" (L4) which checks out given the test case:

```golang
		{
			note:    "functions",
			queries: []string{"data.test.f([1,x])"},
			modules: []string{`  # L1
				package test   # L2
				f([a, b]) {          # L3
					a = b       # L4
				}
			`},
		},
```

However, as you can see, the statements for the query have inherited the location info from the last expr in the function. This is because the call statements are planned AFTER the function definition has been planned.

```
$ go test ./internal/planner/ -v -run TestPlannerHelloWorld/functions
=== RUN   TestPlannerHelloWorld
=== RUN   TestPlannerHelloWorld/functions
*ir.Policy Policy
| *ir.Static Static (2 strings)
| | *ir.StringConst &{Value:x}
| | *ir.StringConst &{Value:1}
| *ir.Plans &{Plans:[Plan test (2 blocks)]}
| | *ir.Plan Plan test (2 blocks)
| | | *ir.Block Block (1 statements)
| | | | *ir.MakeStringStmt <0:0:0>
| | | *ir.Block Block (10 statements)
| | | | *ir.MakeArrayStmt <0:6:4>
| | | | *ir.MakeNumberRefStmt <0:6:4>
| | | | *ir.ArrayAppendStmt <0:6:4>
| | | | *ir.ArrayAppendStmt <0:6:4>
| | | | *ir.CallStmt <0:6:4>
| | | | *ir.MakeBooleanStmt <0:6:4>
| | | | *ir.NotEqualStmt <0:6:4>
| | | | *ir.MakeObjectStmt <0:6:4>
| | | | *ir.ObjectInsertStmt <0:6:4>
| | | | *ir.ResultSetAdd <0:6:4>
| *ir.Funcs Funcs (1 funcs)
| | *ir.Func g0.data.test.f (3 params: [0 1 3], 2 blocks)
| | | *ir.Block Block (14 statements)
| | | | *ir.IsArrayStmt <0:1:1>
| | | | *ir.LenStmt <0:1:1>
| | | | *ir.MakeNumberIntStmt <0:1:1>
| | | | *ir.EqualStmt <0:1:1>
| | | | *ir.MakeNumberIntStmt <0:1:1>
| | | | *ir.AssignIntStmt <0:1:1>
| | | | *ir.DotStmt <0:1:1>
| | | | *ir.AssignVarStmt <0:1:1>
| | | | *ir.AssignIntStmt <0:1:1>
| | | | *ir.DotStmt <0:1:1>
| | | | *ir.AssignVarStmt <0:1:1>
| | | | *ir.EqualStmt <0:6:4>
| | | | *ir.MakeBooleanStmt <0:6:4>
| | | | *ir.AssignVarOnceStmt <0:6:4>
| | | *ir.Block Block (1 statements)
| | | | *ir.ReturnLocalStmt <0:0:0>
--- PASS: TestPlannerHelloWorld (0.00s)
    --- PASS: TestPlannerHelloWorld/functions (0.00s)
PASS
ok      github.com/open-policy-agent/opa/internal/planner       0.003s
```

Reply:
> I'm assuming 0:1:1 is some empty/default value

Almost. It's the index into the files addresses array, row (line) and col. The first "file" we see typically is "<query>", so 0:1:1 would refer to the query. 

This indeed illustrates the problem nicely. My approach was silly, I had tried to trigger a runtime error instead of examining the IR 🤦‍♂️

I'm going to fiddle with the proper places to reset the location state.

Reply:
Another 🤦 I had switched row/col in the `String()` method on location, it's meant to be 4:6, not 6:4.

Reply:
Side show, but before https://github.com/open-policy-agent/opa/pull/2960/commits/8f1492b3896af6a552913b68f44f6b510b29238e we'd always thought our source locations were `<query>` because the filename set by `ast.MustParseModule` is `""`.

Reply:
Resetting before calling `iter` (https://github.com/open-policy-agent/opa/pull/2960/commits/788ad9fb4e2bcbf424cef0c2978260f6f7dd16a4) got is to this:

```
*ir.Policy Policy
| *ir.Static Static (2 strings)
| | *ir.StringConst &{Value:x}
| | *ir.StringConst &{Value:1}
| *ir.Plans &{Plans:[Plan test (2 blocks)]}
| | *ir.Plan Plan test (2 blocks)
| | | *ir.Block Block (1 statements)
| | | | *ir.MakeStringStmt <0:0:0>
| | | *ir.Block Block (10 statements)
| | | | *ir.MakeArrayStmt <0:1:1>
| | | | *ir.MakeNumberRefStmt <0:1:1>
| | | | *ir.ArrayAppendStmt <0:1:1>
| | | | *ir.ArrayAppendStmt <0:1:1>
| | | | *ir.CallStmt <0:1:1>
| | | | *ir.MakeBooleanStmt <0:1:1>
| | | | *ir.NotEqualStmt <0:1:1>
| | | | *ir.MakeObjectStmt <0:0:0>
| | | | *ir.ObjectInsertStmt <0:0:0>
| | | | *ir.ResultSetAdd <0:0:0>
| *ir.Funcs Funcs (1 funcs)
| | *ir.Func g0.data.test.f (3 params: [0 1 3], 2 blocks)
| | | *ir.Block Block (14 statements)
| | | | *ir.IsArrayStmt <0:1:1>
| | | | *ir.LenStmt <0:1:1>
| | | | *ir.MakeNumberIntStmt <0:1:1>
| | | | *ir.EqualStmt <0:1:1>
| | | | *ir.MakeNumberIntStmt <0:1:1>
| | | | *ir.AssignIntStmt <0:1:1>
| | | | *ir.DotStmt <0:1:1>
| | | | *ir.AssignVarStmt <0:1:1>
| | | | *ir.AssignIntStmt <0:1:1>
| | | | *ir.DotStmt <0:1:1>
| | | | *ir.AssignVarStmt <0:1:1>
| | | | *ir.EqualStmt <1:4:6>
| | | | *ir.MakeBooleanStmt <0:1:1>
| | | | *ir.AssignVarOnceStmt <0:1:1>
| | | *ir.Block Block (1 statements)
| | | | *ir.ReturnLocalStmt <0:0:0>
```
Note that the row/cols are switched compared to before. Also, we not get a true zero value, `<0:0:0>`.